### PR TITLE
diagnostic: add check for broken taps

### PR DESCRIPTION
**Pull request by gromgit**
Detect half-baked core taps that show up on a fairly regular basis (e.g. #11465, https://github.com/Homebrew/brew/issues/11243).

A half-baked tap wouldn't have a complete Git config, and is always somewhere below `HOMEBREW_REPOSITORY`, so any Git operation would refer to the Brew repo instead. We simply need to test for any of:

1. Empty tap origin
2. Empty tap HEAD
3. Tap HEAD == Brew HEAD

- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally? 

-----

### CODE REVIEW

File: Library/Homebrew/diagnostic.rb [hunk 1]
```diff
@@ -147,6 +147,17 @@ def examine_git_origin(repository_path, desired_origin)
         end
       end
 
+      def broken_tap?(tap_instance)
+        brew_instance = HOMEBREW_REPOSITORY
+        return if !Utils::Git.available? || !brew_instance.git?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/diagnostic.rb:152]:
```suggestion
        return false unless Utils::Git.available?
        return false unless HOMEBREW_REPOSITORY.git?
```

File: Library/Homebrew/diagnostic.rb [hunk 2]
```diff
@@ -147,6 +147,17 @@ def examine_git_origin(repository_path, desired_origin)
         end
       end
 
+      def broken_tap?(tap_instance)
+        brew_instance = HOMEBREW_REPOSITORY
```
Comment (MikeMcQuaid) [inline Library/Homebrew/diagnostic.rb:151]:
```suggestion
```

File: Library/Homebrew/diagnostic.rb [hunk 3]
```diff
@@ -147,6 +147,17 @@ def examine_git_origin(repository_path, desired_origin)
...
+        return if !Utils::Git.available? || !brew_instance.git?
+
+        tap_origin = tap_instance.remote
+        tap_head = tap_instance.git_head
+        brew_head = brew_instance.git_head
+
+        tap_origin.nil? || tap_head.nil? || tap_head == brew_head
```
Comment (MikeMcQuaid) [inline Library/Homebrew/diagnostic.rb:158]:
```suggestion
        return true if tap.remote.blank?
        
        tap_head = tap.git_head
        return true if tap_head.blank?
        
        tap_head == HOMEBREW_REPOSITORY.git_head
```

File: Library/Homebrew/diagnostic.rb [hunk 4]
```diff
@@ -147,6 +147,17 @@ def examine_git_origin(repository_path, desired_origin)
         end
       end
 
+      def broken_tap?(tap_instance)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/diagnostic.rb:150]:
```suggestion
      def broken_tap?(tap)
```

File: Library/Homebrew/diagnostic.rb [hunk 5]
```diff
@@ -558,15 +569,34 @@ def check_brew_git_origin
...
+        coretap = CoreTap.instance
+        if broken_tap?(coretap)
+          <<~EOS
+            Homebrew/homebrew-core was not tapped properly.
+            To fix:
+              rm -rf "#{coretap.path}"
+              brew tap homebrew/core
+          EOS
+        else
+          examine_git_origin(CoreTap.instance.path, Homebrew::EnvConfig.core_git_remote)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/diagnostic.rb:582]:
```suggestion
          examine_git_origin(coretap, Homebrew::EnvConfig.core_git_remote)
```

File: Library/Homebrew/diagnostic.rb [hunk 6]
```diff
@@ -558,15 +569,34 @@ def check_brew_git_origin
...
+      def check_coretap_integrity
+        coretap = CoreTap.instance
+        if broken_tap?(coretap)
+          <<~EOS
+            Homebrew/homebrew-core was not tapped properly.
+            To fix:
+              rm -rf "#{coretap.path}"
+              brew tap homebrew/core
+          EOS
+        else
```
Comment (MikeMcQuaid) [inline Library/Homebrew/diagnostic.rb:581]:
```suggestion
          EOS
          return
        end
```

which saves the need for the `else`/indent (and same below)

File: Library/Homebrew/diagnostic.rb [hunk 7]
```diff
@@ -558,15 +569,34 @@ def check_brew_git_origin
...
         default_cask_tap = Tap.default_cask_tap
         return unless default_cask_tap.installed?
 
-        examine_git_origin(default_cask_tap.path, default_cask_tap.remote)
+        if broken_tap?(default_cask_tap)
+          <<~EOS
+            Homebrew/homebrew-cask was not tapped properly.
+            To fix:
+              rm -rf "#{default_cask_tap.path}"
+              brew tap homebrew/cask
```
Comment (MikeMcQuaid) [inline Library/Homebrew/diagnostic.rb:595]:
```suggestion
            #{default_cask_tap.full_name} was not tapped properly.
            To fix:
              rm -rf "#{default_cask_tap.path}"
              brew tap #{default_cask_tap.name}
```

which also makes me wonder if `broken_tap?` (without the `?`) could return this message itself?

**Review by MikeMcQuaid — COMMENTED** [review]
love this idea, nice work so far @gromgit!

**Comment by gromgit** [comment]
@MikeMcQuaid, thanks much for the comprehensive review! I've refactored my changes based on your last comment:
> which also makes me wonder if broken_tap? (without the ?) could return this message itself?

It's certainly more streamlined now. :grin:

File: Library/Homebrew/diagnostic.rb [hunk 8]
```diff
@@ -147,6 +147,27 @@ def examine_git_origin(repository_path, desired_origin)
...
+        return unless Utils::Git.available?
+        return unless HOMEBREW_REPOSITORY.git?
+
+        return broken_tap_msg(tap) if tap.remote.blank?
+
+        tap_head = tap.git_head
+        return broken_tap_msg(tap) if tap_head.blank?
+
+        return broken_tap_msg(tap) if tap_head == HOMEBREW_REPOSITORY.git_head
+      end
```
Comment (MikeMcQuaid) [inline Library/Homebrew/diagnostic.rb:169]:
```suggestion
      def broken_tap(tap)
        return unless Utils::Git.available?
        return unless HOMEBREW_REPOSITORY.git?

        message = <<~EOS
          #{tap.full_name} was not tapped properly! Run:
            rm -rf "#{tap.path}"
            brew tap #{tap.name}
        EOS

        return message if tap.remote.blank?

        tap_head = tap.git_head
        return message if tap_head.blank?
        return if tap_head != HOMEBREW_REPOSITORY.git_head
        
        message
      end
```

Comment (gromgit) [reply Library/Homebrew/diagnostic.rb]: Argh, sometimes I miss the obvious. :rofl:

**Review by MikeMcQuaid — APPROVED** [review]
Great work here as usual @gromgit!

**Pull request merged**